### ISSUE TITLE
Add support for .elpaignore

### ISSUE THREAD

**Issue by minad** [melpa/melpa#8287]
(Non)GNU ELPA uses .elpaignore files to ignore files during the package building. It would be good if MELPA supported that too such that we have a common standard to ignore files. We would avoid duplicating the ignore information. Multiple packages are distributed via both NonGNU ELPA and MELPA. These packages could profit from .elpaignore support.

**Comment by tarsius** [comment]
The approaches taken by (non)GNU Elpa and Melpa differ greatly. (non)GNU Elpa includes "everything that is not definitely undesirable", while Melpa excludes "everything that is not actually useful".

One can argue about what approach is better, but it does not make sense and is dangerous, to use a package-specific rules designed for one approach when actually using the other approach.

**Comment by minad** [comment]
@tarsius Of course. Actually I prefer the approach of MELPA. But this doesn't seem relevant for this issue? I mean all files listed in .elpaignore are intentionally excluded by the package author and most likely the exclusion list should apply to all distribution channels.

**Comment by akirak** [comment]
> all files listed in .elpaignore are intentionally excluded by the package author and most likely the exclusion list should apply to all distribution channels

This doesn't seem to be the case. Documentation source files (e.g. `*.texi`, `*.org`, etc.) are usually ignored in ELPA:

- https://github.com/renzhengeek/dotfiles/blob/f0b54ed280ec2c46cd45821377679e975ee402fb/Condor-Emacs/.emacs.d/elpa/bind-key-2.4.1/.elpaignore
- https://github.com/gobbletown/shane-emacs-config-dump/blob/ecbab875a369c8afec18e89178743729717ca92e/packages28/web-server-0.1.2/.elpaignore
- https://github.com/gobbletown/shane-emacs-config-dump/blob/ecbab875a369c8afec18e89178743729717ca92e/packages28/ivy-0.13.4/.elpaignore

while MELPA *supports* `*.texi` files via [its recipe spec](https://github.com/melpa/melpa#recipe-format).

Furthermore, MELPA copies files to the root directory of the output (like `doc/*.texi` to `*.texi`), so applying `.elpaignore` before/after the existing build step wouldn't work.

**Comment by minad** [comment]
Okay I understand. That's too bad.

**Comment by tarsius** [comment]
> I mean all files listed in .elpaignore are intentionally excluded by the package author and most likely the exclusion list should apply to all distribution channels.

Sure, but they are written with GNU Elpa in mind.

And more importantly Melpa currently has the nice property that the rules can be found in one place (in the respective recipe in the melpa repository). If we accept other sources, then we always have to look at all potential sources to get the full picture.

Using `.elpaignore` might help in some case when the package author forgets to update the Melpa recipe, *but it would always better if they also updated the Melpa recipe*.

In other cased (such as just mentioned by @akirak) the values in that file are wrong for Melpa, so we would have to filter out those rules. And that would mean figuring out which rules have to be filtered out and how to do it. Once `.elpaignore` is respected, but only partially, it becomes very adventurous to use it all, because nobody will always have all the rules and exceptions in their mind, while writing `.elpaignore`.

**Comment by minad** [comment]
It would work if we could specify some MELPA/ELPA specific groups in the .elpaignore file, e.g., like in toml/ini files. But it would certainly come with some adoption pain and it would require convincing the ELPA maintainers. 

I also agree with @tarsius that it would be nice to maintain all information at a single place. The ELPA maintainers insist that this information should be maintained in .elpaignore. Therefore this issue in the hope to harmonize the situation.

Overall the incoherent situation is a bit inconvenient for package maintainers if packages live on both ELPA/MELPA. See also the versioning problem #8135.

**Comment by minad** [comment]
I just learned about this now - the default Emacs package manager package.el understands .elpaignore itself since Emacs 29. I think supporting .elpaignore in MELPA should be reconsidered.

https://github.com/emacs-mirror/emacs/blob/f2b2c752a5920c1152fd70d917e0fc5a3c9728c4/lisp/emacs-lisp/package.el#L613-L630

**Comment by minad** [comment]
As a backward compatible solution, a directive `:ignorefile` would be possible, in addition to the existing `:exclude` directive:


~~~ emacs-lisp
(somepkg
 :fetcher github
 :repo "emacsdeveloper/somepkg"
 :files (:defaults (:ignorefile ".elpaignore")))
~~~

**Comment by minad** [comment]
@tarsius A few years later - what is your opinion about this feature now? Would it be acceptable as an opt-in feature? I prefer to use .elpaignore instead of the MELPA ignore mechanism, since this way the ignore information does not have to scattered over three repositories (the project repository, MELPA repository, ELPA repository).

**Comment by minad** [comment]
Maybe an opt-in addition is not worth the complexity, and I don't know how widely .elpaignore is used. But I think it might be a preferable mechanism for many package authors, since they don't have to change the MELPA recipe to apply changes.

**Comment by tarsius** [comment]
I still think this is a bad idea and I don't want to do it. Please give up on this idea.

**Comment by minad** [comment]
Frankly I don't see how it is a bad idea if it is opt-in - this would solve all compatibility issues. The ELPA guys think it is a good idea, therefore they implemented it, and they even argue that one should rather list excluded files in .elpaignore instead of maintaining the information. However I understand that you don't want to implement it, or don't want to maintain it. That's okay.

**Comment by tarsius** [comment]
> Frankly I don't see how it is a bad idea if it is opt-in - this would solve all compatibility issues. The ELPA guys think it is a good idea, therefore they implemented it,

I didn't say the approach they took was a bad idea.

> The approaches taken by (non)GNU Elpa and Melpa differ greatly. (non)GNU Elpa includes "everything that is not definitely undesirable", while Melpa excludes "everything that is not actually useful".
>
> One can argue about what approach is better, [...]

But that mixing them is:

> [...] but it does not make sense and is dangerous, to use package-specific rules designed for one approach when actually using the other approach.

**Comment by minad** [comment]
@tarsius After checking my packages again, I think you are right. There is no point in supporting .elpaignore in MELPA. The packages are already as lightweight as possible, since they only include the *.el files. It would only make sense if one switches the approach to include everything in the way ELPA does it (e.g. if .elpaignore is present), but this would go too far, given that MELPA works well as is. I think my only (unrelated) gripe with the MELPA way is that I miss the *.info files, since they are not automatically created from *.org files, but I think I can live with this limitation.

**Comment by minad** [comment]
I should also add that my main motivation is to remove unnecessary data from package builds, and this is already perfectly realized by MELPA. However thanks to this discussion I learned that one of my packages on MELPA is faulty, since it will miss an auxiliary file, which needs to be part of the package. I'll open a PR which updates the recipe.

**Comment by tarsius** [comment]
The "`*.info` is missing" issue also came up in one of the other issues you bumped (though in that case it was about `*.texi`. The way way `package-build` currently handles this is "if we build an 'end-user' format of some documentation, then the 'source' file isn't needed anymore, so we don't ship it". Apparently some (though not very many) package maintainers don't like that.

I am not sure it is worth adding support for keeping the source files, but if we did, it would probably be through some new `:keep-documentation-sources` recipe keyword.

**Comment by minad** [comment]
> The "*.info is missing" issue also came up in one of the other issues you bumped (though in that case it was about *.texi. The way way package-build currently handles this is "if we build an 'end-user' format of some documentation, then the 'source' file isn't needed anymore, so we don't ship it". Apparently some (though not very many) package maintainers don't like that.

I'd agree with that approach, but my package repositories lack the *.texi files, which would have to be created from the *.org files. The ideal scenario for me is to have a clean repository and a clean package build. So ideally I could delete the .elpaignore and wouldn't need *.texi files.
